src/Features/Core/Completion/ICompletionService.cs

```diff
@@ -51,6 +46,25 @@ internal interface ICompletionService : ILanguageService
         Task<string> GetSnippetExpansionNoteForCompletionItemAsync(CompletionItem completionItem, Workspace workspace);
 
         /// <summary>
+        /// Called when a completion item is committed.  The completion service can then use this to
+        /// help inform future completion requests.  For example, it can use this information to 
+        /// help decide if this item is 'better' than another when queried.
+        /// </summary>
+        void CompletionItemComitted(CompletionItem item);
+
+        /// <summary>
+        /// Returns true if 'item1' is a better match for the filter text typed so far versus 
+        /// 'item2'.  The better match will be selected after text is typed.
+        /// </summary>
+        bool IsBetterFilterMatch(CompletionItem item1, CompletionItem item2);
+
+        /// <summary>
+        /// Used to allow the completion service to indicate if this item should be preselected.
+        /// Only called when no filter text has been provided.
+        /// </summary>
+        bool ShouldPreselect(CompletionItem item);
```

Comment:
:bulb: Consider a name like `CanPreselect` or `AllowsPreselection`, since the actual selection is controlled by all of the following:

1. The number of items which are eligible for preselection.
2. Heuristics regarding match strength.
3. In-IDE preferences for whether or not an item is ever initially selected (I forget the name of the particular mode).